File: src/video/sequence.go
package video

import (
	"log"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"sync"

	"github.com/joachimbbp/spritefire/src/mosaic"
	"github.com/joachimbbp/spritefire/src/search"
	"github.com/joachimbbp/spritefire/src/util"
)

func Sequence(sequencePath string, spriteColorDbPath string, spriteSizeIndex int) {
	prefix := strconv.Itoa((util.ResizeResolutions[spriteSizeIndex])) + "_"

	frames, err := os.ReadDir(sequencePath)
	if err != nil {
		log.Fatal(err)
	}

	sequenceData := make(map[string][]util.IndexedSprite)
	spriteColorDb := util.DecodeColorDatabase(spriteColorDbPath)
	tree := search.BuildSearchTree(spriteColorDb)

	var wg sync.WaitGroup
	var mu sync.Mutex

	for _, frame := range frames {
		if filepath.Ext(frame.Name()) != ".png" {
			continue
		}
		wg.Add(1)
		go func(frame os.DirEntry) {
			defer wg.Done()
			framePath := filepath.Join(sequencePath, frame.Name())
			spriteCanvas := mosaic.Canvas(framePath, spriteColorDb, spriteSizeIndex, tree)
			mu.Lock()
			sequenceData[frame.Name()] = spriteCanvas

			mu.Unlock()
		}(frame)
	}
	wg.Wait()

	keys := make([]string, 0, len(sequenceData))
	for k := range sequenceData {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	for _, frameName := range keys {
		mosaic.Draw(sequenceData[frameName], prefix, frameName, spriteSizeIndex)

	}

}

File: cpp/spritefire/src/video/sequence.cpp
#include "sequence.h"

#include <filesystem>
#include <string>

#include "utils/paths.h"
#include "utils/resolution.h"

void drawSequence(const std::string& sequencePath, const std::string& spriteColorDbPath, int spriteSizeIndex) {
    std::filesystem::directory_iterator end;
    for (std::filesystem::directory_iterator it(sequencePath); it != end; it++) {
        if (it->is_regular_file() && it->path().extension() == ".png") {
            std::string framePath = it->path().string();
            // Canvas canvas = Canvas(framePath, DATABASE_PATH, RESIZE_RESOLUTIONS[spriteSizeIndex]);
            // mosaic.drawCanvas(canvas, framePath.filename(), RESIZE_RESOLUTIONS[spriteSizeIndex]);
        }
    }
}